crates/core/tedge/src/cli/upload/mod.rs

```diff
@@ -0,0 +1,123 @@
+use crate::command::BuildCommand;
+use crate::command::BuildContext;
+use crate::command::Command;
+use crate::ConfigError;
+use c8y_api::http_proxy::C8yEndPoint;
+use std::collections::HashMap;
+use std::path::PathBuf;
+use tedge_config::ProfileName;
+
+mod c8y;
+
+#[derive(clap::Subcommand, Debug)]
+pub enum UploadCmd {
+    /// Upload a file to Cumulocity
+    ///
+    /// The command create a new event for the device,
+    /// attach the given file content to this new event,
+    /// and return the event ID.
+    C8y {
+        /// Path to the uploaded file
+        #[clap(long)]
+        file: PathBuf,
```

Comment:
Even though that file path is mandatory argument, I'm assuming that it is accepted as an option rather than a direct argument of the `c8y` subcommand to expand the upload capability to enable uploading to other C8y repositories like software repo, firmware repo etc with future types like `--software`, `--firmware` etc.

Reply:
> Since the file path is mandatory argument, I thought it would be taken as a direct argument of the c8y subcommand. But, I'm assuming that it is accepted as a --file option rather than a direct argument to expand the upload capability to enable uploading to other C8y repositories like software repo, firmware repo etc with future options like --software, --firmware etc.

I'm pretty sure this is even possible to implement as the device user has restricted permissions. Generally device users can only change data that they own, so uploading to the software/firmware repo won't work successfully as it needs to edit an existing managed object by adding a relationship to the uploaded binary (which is also stored as a binary managed object).

I would avoid supporting uploading directly to the software/firmware repo anyway, as these binaries are stored as [managed object binaries](https://cumulocity.com/api/core/#operation/postBinariesCollectionResource) which aren't cleaned up by the Cumulocity data retention feature. Plus device's shouldn't be responsible for publishing this data directly, as such a binary needs to be validated by a user (or process) before promoting it from one device to other device (imagine if a device uploads a software binary which is in armhf CPU architecture, and another device which has arm64 tries to install it).